src/ws/session.c

```diff
@@ -515,12 +515,13 @@ open_session (pam_handle_t *pamh)
       return res;
     }
 
-  /* Yes, buf "leaks" */
-  buf = malloc (sizeof (struct passwd) + 8192);
-  if (buf == NULL)
+  if (!pwd_buf)
+    pwd_buf = malloc (sizeof (struct passwd) + 8192);
+
+  if (pwd_buf == NULL)
     res = ENOMEM;
   else
-    res = getpwnam_r (name, buf, (char *)(buf + 1), 8192, &pwd);
+    res = getpwnam_r (name, pwd_buf, (char *)(pwd_buf + 1), 8192, &pwd);
```

Comment:
I'll look at making the buffer static, but we can't just change pwd - it's global, used all over the file and [`getpwnam_r`](https://linux.die.net/man/3/getpwnam_r) can also point this to a different static area.